bee-protocol/src/tangle/mod.rs

```diff
@@ -42,6 +49,7 @@ pub struct MsTangle {
     snapshot_index: AtomicU32,
     pruning_index: AtomicU32,
     entry_point_index: AtomicU32,
+    tip_pool: Arc<RwLock<WurtsTipPool>>,
```

Comment:
why `RwLock` instead of a `DashSet` for the tip pool?

Reply:
For example when the scores of the tips are getting updated (on an incoming milestone) but the user performs tip-selection at the very same time, this could lead to the situation where this tip selection is performed still on old values. The RwLock helps helps to synchronize the functions.

Reply:
Isn't that exactly what DashSet does though ?

Reply:
Isn't it possible that it that scores being updated but the user at the same time performs TSA and gets a lazy tip? If he would have waited a bit longer till the scores were updated he would got a non-lazy tip

Reply:
@christianpoveda can you remember our discussion? if we want that locked behaviour between the functions of the tip-pool, do you think this is fine?

Reply:
No, from my understanding updating can only make a former non-lazy tip turn into a now lazy tip, but you cannot make an already lazy tip non-lazy through an update.

Reply:
For me the choice between a `Mutex` and an `RwLock` depends on how often it is written to/read from the protected inner data structure. In case of a tip pool I find it more likely that both operations happen regularly (in case of a well used node). If I am not wrong, that an RwLock optimizes for situations in which you have many more read operations than write operations, so I tend to use a `Mutex` here, or just go with a `DashSet`.

Reply:
by just looking at the API I assume is possible to mutate the tip pool from different threads at the same time and potentially cause a deadlock if dashmap/set were being used. How likely is that to happen is not something I can tell right now.

About using Mutex instead RwLock, I agree with Alex, there's only one method reading from the lock, all the other ones are mutable borrows. So unless the read-only method is called more frequently I think it'd be better to use a Mutex.

To be honest I'd rather avoid using DashMap/Set as this has caused a lot of trouble in the past that we weren't able to actually solve (the last deadlock stopped happening during the tokio drift, we don't know why). If this isn't a hot path I'd be tempted to do message passing even to have a lock-free solution. 

Other aspect that we haven't considered here is if I'd be better to use an async-aware mutex instead of the `std` one. However I think this is not so relevant right now.

Reply:
With normal node usage, the read-only method is only called from time to time. So, agree with you guys, `Mutex` should suit it better than `RwLock`. So at least we should use the `std` `Mutex`.